### ISSUE TITLE
Connection [FAIL] Reason: unknown on trying to connect to peer

### ISSUE THREAD

**Issue by fridgecow** [albfan/miraclecast#250]
I can't seem to connect to any peer, either as a sink or as just a peer.

Attached are my logs, generated by the method from the FAQ

Sink device is Surface Pro 4, Source device is Nexus 4 (as you can see). Both support P2P, as I can cast to other miracast devices and act as a sink in Windows.

[logs.zip](https://github.com/albfan/miraclecast/files/1579293/logs.zip)

**Comment by hardiksingh-rathore** [comment]
Is it possible to provide output of iw list command ?

**Comment by fridgecow** [comment]
Of course:

[iw list.txt](https://github.com/albfan/miraclecast/files/1579348/iw.list.txt)

**Comment by hardiksingh-rathore** [comment]
Ok  this one from Nexus 4?

**Comment by fridgecow** [comment]
The iw list is from the Surface Pro 4, my apologies. The connection that is
trying to be established (to the SP4) is from a Nexus 4 - my apologies that
there's other noise in there, it's my neighbour's TV.

**Comment by hardiksingh-rathore** [comment]
I haven't gone through complete logs yet but I hope you followed all steps suggested by Alberto.
Like: $ $ sudo kill -9 $(ps -ef | grep wpa_supplican[t] | awk '{print $2}') or run $ kill-wpa.sh before starting miraclecast application.

I can see some failure logs 
TRACE: wpa: raw message: IFNAME=p2p-wlp2s0-0 <3>Failed to start AP functionality

So have you enabled  CONFIG_AP option in wpa_supplicant ?
If not, you can follow the guide as given below
https://thangamaniarun.wordpress.com/2013/03/03/how-to-use-wi-fi-direct-on-androidubuntu-part1/

**Comment by hardiksingh-rathore** [comment]
Link given above will not help in wpa_supplicant configuration
So you can follow 
http://www.linuxfromscratch.org/blfs/view/svn/basicnet/wpa_supplicant.html  
or
http://riya80211.blogspot.in/2013/07/p2p-configuration-of-wpasupplicant-in.html
or 
you can share you wpa_supplicant .config file.

**Comment by fridgecow** [comment]
So, I will need to recompile wpa_supplicant? As it stands I'm using the
default Ubuntu wpa_supplicant package, I have no idea what
flags/configuration it was built with. Is there a good place to find out?

**Comment by hardiksingh-rathore** [comment]
I am sharing my wpa_supplicant-2.6 package with .config ( Install it on your own risk !)
$ tar -xvf wpa_supplicant-2.6.tar.xz
$cd wpa_supplicant-2.6/wpa_supplicant
$ make
$ sudo install.sh ( Warning: it's working on my ubuntu-16.04 system but for this "sudo" you have to take responsibility)

[wpa_supplicant-2.6.zip](https://github.com/albfan/miraclecast/files/1579524/wpa_supplicant-2.6.zip)

**Comment by fridgecow** [comment]
I've recompiled wpa_supplicant from your files and running "wpa_supplicant"
now tells me it's v2.6 (it was 2.4 before). Great!

However, miracast sink (or, in general, connecting as a peer, tested with
miracle_wifictl) still does not work - providing the same output as before.

Thank you for your help so far!

**Comment by hardiksingh-rathore** [comment]
Ok. Whenever you need any support on this, Let us know the steps (command !) you are following
and provide the logs as suggested in FAQ.

**Comment by fridgecow** [comment]
Alright. Ubuntu 17.10, Gnome3 (with X.org because I thought that might be causing an issue)

Everything default except for, of course, the wpa_supplicant we just installed.

Commands I run:
```
$ sudo kill -9 $(ps -ef | grep wpa_supplican[t] | awk '{print $2}')
$ sudo journalctl -f |& tee journal.log #Lots of output
#In a new terminal
$ sudo miracle-wifid --log-level trace |& tee wifid.log #Lots of output
#Again, in a new terminal
$ sudo miracle-sinkctl --log-level trace --log-journal-level trace |& tee sink.log

#Within sinkctl
[ADD] Link: 4
[miraclectl] # run 4
now running on link 4
[ADD] Peer: 9e:ad:97:62:02:0e@4
[ADD] Peer: 12:68:3f:7c:8b:50@4
[REMOVE] Peer: 9e:ad:97:62:02:0e@4
[REMOVE] Peer: 12:68:3f:7c:8b:50@4
[ADD] Peer: 12:68:3f:7c:8b:50@4
[PROV] Peer: 12:68:3f:7c:8b:50@4 Type: pbc PIN: 
[PROV] Peer: 12:68:3f:7c:8b:50@4 Type: pbc PIN: 
[PROV] Peer: 12:68:3f:7c:8b:50@4 Type: pbc PIN: 
[GO NEG] Peer: 12:68:3f:7c:8b:50@4 Type: pbc PIN: 
[FAIL] Peer: 12:68:3f:7c:8b:50@4 Reason: unknown
[ADD] Peer: 9e:ad:97:62:02:0e@4
[PROV] Peer: 12:68:3f:7c:8b:50@4 Type: pbc PIN: 
[GO NEG] Peer: 12:68:3f:7c:8b:50@4 Type: pbc PIN: 
[FAIL] Peer: 12:68:3f:7c:8b:50@4 Reason: unknown
[miraclectl] # quit
ERROR: cannot change p2p-scanning state on link 4 to 0: Transport endpoint is not connected (ctl_link_set_p2p_scanning() in /home/tom/Documents/miraclecast/src/ctl/ctl-wifi.c:723)
no longer waiting for peer Android_9157 (12:68:3f:7c:8b:50@4)
no longer running on link 4

```

**Comment by albfan** [comment]
Remember there's a chat on https://gitter.im/albfan/miraclecast to refine issue, all this is a little chaotic.

@fridgecow try the `--use-dev` option with miracle-wifid. WPA not always resend all the parameters and that confuses miraclecast

**Comment by fridgecow** [comment]
I didn't realise there was a chat! I'll use that from now on - but I'll
post any solutions we find back here, lest we forget xkcd 979!

**Comment by albfan** [comment]
From your code, wpa_supplicant is failing here (I grab first wpa_supplicant code I found)

- wifid.log:
```
DEBUG: supplicant: unhandled wpas-event: IFNAME=p2p-wlp2s0-0 <3>Failed to start AP functionality (supplicant_event() in /home/tom/Documents/miraclecast/src/wifi/wifid-supplicant.c:1507)
```
https://github.com/realdesktop/wpa_supplicant/blob/a4d9a1df6175fd9c290b06770b0dc9f0af0bdfc8/wpa_supplicant/ap.c#L377

- journal.log
```
Dec 21 17:03:21 tom-SP4 miracle-wifid-wlp2s0-4[6045]: P2P-GO-NEG-REQUEST 12:68:3f:7c:8b:50 dev_passwd_id=4 go_intent=0
Dec 21 17:03:21 tom-SP4 miracle-wifid-wlp2s0-4[6045]: P2P-FIND-STOPPED
Dec 21 17:03:21 tom-SP4 kernel: mwifiex_pcie 0000:02:00.0: Can not set Tx data rate in disconnected state
Dec 21 17:03:21 tom-SP4 NetworkManager[5665]: <info>  [1513875801.1416] wifi-nl80211: (p2p-wlp2s0-1): using nl80211 for WiFi device control
Dec 21 17:03:21 tom-SP4 systemd-udevd[6297]: link_config: autonegotiation is unset or enabled, the speed and duplex are not writable.
Dec 21 17:03:21 tom-SP4 NetworkManager[5665]: <info>  [1513875801.1816] devices added (path: /sys/devices/pci0000:00/0000:00:1d.0/0000:02:00.0/net/p2p-wlp2s0-1, iface: p2p-wlp2s0-1)
Dec 21 17:03:21 tom-SP4 NetworkManager[5665]: <info>  [1513875801.1817] device added (path: /sys/devices/pci0000:00/0000:00:1d.0/0000:02:00.0/net/p2p-wlp2s0-1, iface: p2p-wlp2s0-1): no ifupdown configuration found.
Dec 21 17:03:23 tom-SP4 ModemManager[673]: <info>  Couldn't check support for device at '/sys/devices/pci0000:00/0000:00:1d.0/0000:02:00.0': not supported by any plugin
Dec 21 17:03:25 tom-SP4 miracle-wifid-wlp2s0-4[6045]: P2P-GO-NEG-SUCCESS role=GO freq=2412 ht40=0 peer_dev=12:68:3f:7c:8b:50 peer_iface=12:68:3f:7c:0b:50 wps_method=PBC
Dec 21 17:03:25 tom-SP4 kernel: mwifiex_pcie 0000:02:00.0: Can not set Tx data rate in disconnected state
Dec 21 17:03:25 tom-SP4 kernel: IPv6: ADDRCONF(NETDEV_UP): p2p-wlp2s0-1: link is not ready
Dec 21 17:03:25 tom-SP4 kernel: mwifiex_pcie 0000:02:00.0: overriding ps_mode to false
Dec 21 17:03:25 tom-SP4 kernel: mwifiex_pcie 0000:02:00.0: Can not set Tx data rate in disconnected state
Dec 21 17:03:25 tom-SP4 miracle-wifid-wlp2s0-4[6045]: p2p-wlp2s0-1: Failed to start AP functionality
Dec 21 17:03:25 tom-SP4 miracle-wifid-wlp2s0-4[6045]: P2P-GROUP-FORMATION-FAILURE
Dec 21 17:03:25 tom-SP4 miracle-wifid-wlp2s0-4[6045]: P2P-GROUP-REMOVED p2p-wlp2s0-1 GO reason=FORMATION_FAILED
Dec 21 17:03:25 tom-SP4 kernel: ------------[ cut here ]------------
Dec 21 17:03:25 tom-SP4 kernel: WARNING: CPU: 0 PID: 6045 at net/wireless/util.c:1049 cfg80211_change_iface+0x1a8/0x440 [cfg80211]
Dec 21 17:03:25 tom-SP4 kernel: Modules linked in: rfcomm bnep btusb btrtl btbcm btintel bluetooth ecdh_generic xt_CHECKSUM iptable_mangle ipt_MASQUERADE nf_nat_masquerade_ipv4 iptable_nat nf_nat_ipv4 nf_nat nf_conntrack_ipv4 nf_defrag_ipv4 xt_conntrack nf_conntrack libcrc32c ipt_REJECT nf_reject_ipv4 xt_tcpudp bridge stp llc ebtable_filter ebtables ip6table_filter ip6_tables iptable_filter hid_sensor_rotation hid_sensor_accel_3d hid_sensor_als hid_sensor_gyro_3d hid_sensor_trigger industrialio_triggered_buffer kfifo_buf hid_sensor_iio_common industrialio binfmt_misc joydev intel_ipts nls_iso8859_1 hid_sensor_hub input_leds snd_hda_codec_hdmi snd_soc_skl snd_soc_skl_ipc snd_soc_sst_ipc snd_soc_sst_dsp snd_hda_ext_core snd_hda_codec_realtek snd_soc_sst_match snd_hda_codec_generic snd_soc_core snd_compress ac97_bus snd_pcm_dmaengine
Dec 21 17:03:25 tom-SP4 kernel:  snd_hda_intel intel_rapl snd_hda_codec x86_pkg_temp_thermal snd_hda_core intel_powerclamp snd_hwdep coretemp snd_pcm kvm_intel kvm mwifiex_pcie snd_seq_midi snd_seq_midi_event irqbypass crct10dif_pclmul mwifiex crc32_pclmul snd_rawmidi ghash_clmulni_intel pcbc cfg80211 snd_seq snd_seq_device snd_timer aesni_intel aes_x86_64 crypto_simd glue_helper cryptd hid_multitouch intel_cstate intel_rapl_perf snd soundcore ov5693(C) v4l2_common videodev idma64 intel_pch_thermal virt_dma mei_me media shpchp mei intel_lpss_pci soc_button_array surfacepro3_button mac_hid intel_lpss_acpi intel_lpss tpm_crb acpi_pad parport_pc ppdev lp parport ip_tables x_tables autofs4 uas usb_storage i915 i2c_algo_bit drm_kms_helper syscopyarea sysfillrect sysimgblt fb_sys_fops nvme drm nvme_core i2c_hid usbhid hid
```

but I can't tell you why wpa_supplicant fails on that with your device

What's your environment (distro, version, wpa_supplicant version, wireless card, kernel, drivers)?

**Comment by fridgecow** [comment]
In order:

Ubuntu 17.10, wpa_supplicant v2.6 (built from above), Marvell AVASTAR Wireless-AC, 4.14.4-jakeday-surface (4.14.4 with some custom patches for touchscreen etc), mwifiex_pcie

**Comment by fridgecow** [comment]
The problem seems to stem from the line `Can not set Tx data rate in disconnected state`

Fortunately, someone has had this issue before. [Unfortunately, nobody ever replied to them with support.](https://www.spinics.net/lists/linux-wireless/msg162039.html).

I'm closing this issue since it's not a bug for you guys. Thanks for all your help!

**Comment by albfan** [comment]
Anyway if you feel brave enough to research on that plumbings I'm alway interested, but that would imply clone that kernel, those drivers, wpa_supplicant and link here a different issue

Warning: Is a hard and dangerous path

**Comment by albfan** [comment]
https://github.com/torvalds/linux/blob/ead68f216110170ec729e2c4dec0aad6d38259d7/drivers/net/wireless/marvell/mwifiex/cfg80211.c#L1649-L1653

> It's a dangerous business, Frodo, going out your door. You step onto the road, and if you don't keep your feet, there's no knowing where you might be swept off to.

**Comment by fridgecow** [comment]
Haha! I'm afraid I'm out of my depth. I tried removing this check to see if
the chipset does indeed support changing Tx rates when disconnected -
however, sinkctl fails at the same point (just without the useful error
message).

I can open up a new issue to track any progress on this and post my new
logs, but I don't think I'll be able to help bring it forward.
